fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/SetPreAggStatus.java

```diff
@@ -0,0 +1,579 @@
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+
+/**
+ * SetPreAggStatus
+ * top-down tranverse the plan tree and collect required info into PreAggValidateContext
+ * when get to the bottom LogicalOlapScan node, we set the preagg status using info in PreAggValidateContext
+ */
+public class SetPreAggStatus extends DefaultPlanRewriter<Stack<SetPreAggStatus.PreAggValidateContext>>
+        implements CustomRewriter {
+
+    /**
+     * PreAggValidateContext
+     */
+    public static class PreAggValidateContext {
+        public List<Expression> filterConjuncts = new ArrayList<>();
+        public List<Expression> joinConjuncts = new ArrayList<>();
+        public List<Expression> groupByExpresssions = new ArrayList<>();
+        public Set<AggregateFunction> aggregateFunctions = new HashSet<>();
+
+        void addJoinInfo(LogicalJoin logicalJoin) {
+            joinConjuncts.addAll(logicalJoin.getExpressions());
+        }
+
+        void addFilterConjuncts(List<Expression> conjuncts) {
+            filterConjuncts.addAll(conjuncts);
+        }
+
+        void addGroupByExpresssions(List<Expression> expressions) {
+            groupByExpresssions.addAll(expressions);
+        }
+
+        void addAggregateFunctions(Set<AggregateFunction> functions) {
+            aggregateFunctions.addAll(functions);
+        }
+
+        void replaceExpressions(Map<Slot, Expression> replaceMap) {
+            if (!filterConjuncts.isEmpty()) {
+                filterConjuncts = Lists.newArrayList(ExpressionUtils.replace(filterConjuncts, replaceMap));
+            }
+            if (!joinConjuncts.isEmpty()) {
+                joinConjuncts = Lists.newArrayList(ExpressionUtils.replace(joinConjuncts, replaceMap));
+            }
+            if (!groupByExpresssions.isEmpty()) {
+                groupByExpresssions = Lists.newArrayList(ExpressionUtils.replace(groupByExpresssions, replaceMap));
+            }
+            if (!aggregateFunctions.isEmpty()) {
+                Set<AggregateFunction> newAggregateFunctions = Sets.newHashSet();
+                for (AggregateFunction aggregateFunction : aggregateFunctions) {
+                    newAggregateFunctions
+                            .add((AggregateFunction) ExpressionUtils.replace(aggregateFunction, replaceMap));
+                }
+                aggregateFunctions = newAggregateFunctions;
+            }
+        }
+    }
+
+    @Override
+    public Plan rewriteRoot(Plan plan, JobContext jobContext) {
+        return plan.accept(this, new Stack<>());
+    }
+
+    @Override
+    public Plan visit(Plan plan, Stack<PreAggValidateContext> context) {
+        context.clear();
+        return super.visit(plan, context);
+    }
+
+    @Override
+    public Plan visitLogicalOlapScan(LogicalOlapScan logicalOlapScan, Stack<PreAggValidateContext> context) {
+        if (logicalOlapScan.isPreAggStatusUnSet()) {
```

Comment:
could we check this first to avoid collect info to a already processed table. and add short-circuit way to process dup table and mow table